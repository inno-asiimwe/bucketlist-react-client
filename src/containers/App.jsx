import React, { Component } from 'react';


const App = () => (
  <div>
    <h3>Home</h3>
  </div>
);


export default App;
