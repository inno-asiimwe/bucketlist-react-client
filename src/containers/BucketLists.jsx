/** Component displays all the bucketlists for a given user */
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import _ from 'lodash';
import { clearMessages } from '../actions/action_auth';
import { getBucketlists, deleteBucketlist, searchBucketlists } from '../actions/action_bucketlist';
import Paginate from '../components/paginate';
import SearchBar from '../containers/SearchBar';

export class BucketLists extends Component {
  constructor(props) {
    super(props);
    this.renderBucketlists = this.renderBucketlists.bind(this);
    this.onDelete = this.onDelete.bind(this);
    this.notify = this.notify.bind(this);
    this.toNextPage = this.toNextPage.bind(this);
    this.toPrevPage = this.toPrevPage.bind(this);
  }

  componentDidMount() {
    // Fetch bucketlists for the current page, initial page is 1.
    this.props.getBucketlists(this.props.current);
  }

  /**
   * Function called on click of the delete button.
   * @param {number} id - id of the selected bucketlist.
   */
  onDelete(id) {
    // Dispatch action to delete the selected bucketlist.
    this.props.deleteBucketlist(id, () => this.props.getBucketlists(this.props.current));
  }
  
  /**
   * Function toasts appropriate message to the user.
   */
  notify = (msg) => {
    toast(msg);
  }

  /**
   * Function fetches the next page.
   */
  toNextPage() {
    // Dispatch action to fetch the next page of bucketlists from the API.
    this.props.getBucketlists(this.props.next);
  }

  /**
   * Function fetches the previous page.
   */
  toPrevPage() {
    // Dispatch action to fetch the previous page of bucketlists from API.
    this.props.getBucketlists(this.props.prev);
  }

  /**
   * Function performs a serch for bucketlists with a given term in the name.
   * @param {string} term - term to be searched for.
   */
  bucketlistSearch(term){
    // Dispatch action to search at API level.
    this.props.searchBucketlists(term);
  }

  /**
   * Functions laysout the returned bucketlists in a tabular manner.
   */
  renderBucketlists() {
    return _.map(this.props.bucketlists, bucketlist => (
      <tr>
        <td>
          <Link className="text-xs-left" to={`/bucketlists/${bucketlist.id}`}>
            {bucketlist.name}
          </Link>
        </td>
        <td>
          <Link className="btn btn-success" to={`/bucketlists/${bucketlist.id}`}>
                OPEN
          </Link>
        </td>
        <td>
          <Link className="btn btn-primary" to={`/bucketlists/${bucketlist.id}/edit`}>
                EDIT
          </Link>
        </td>
        <td>
          <button className="btn btn-danger" onClick={() => this.onDelete(bucketlist.id)} >
                DELETE
          </button>
        </td>
      </tr>
    ));
  }

  render() {
    // Toast a success message on successful login
    if (this.props.auth.success) {
      this.notify(this.props.auth.success_msg);
      this.props.clearMessages();
    }
    
    // Delay Search function by 300 milliseconds to avoid a search on every key stroke.
    const bucketlistSearch = _.debounce((term) => { this.bucketlistSearch(term)}, 300)

    return (
      <div className="container">
        <ToastContainer />
        <div>
          <SearchBar onSearchTermChange={bucketlistSearch} />
        </div>
        <div className="jumbotron">
          <div className="text-right">
            <Link className="btn btn-primary" to="/bucketlists/new">
              Add a bucketlist
            </Link>
          </div>
          <div className="text-left">
          </div>
          <h3>Bucketlists table</h3>
          {this.props.pages == 0 && 
            <div> No Bucketlists found</div>
          }
          <table className="table table-bordered">
            <tbody>
              {this.renderBucketlists()}
            </tbody>
          </table>
          <Paginate 
            page={this.props.current}
            pages={this.props.pages}
            onNext={this.toNextPage}
            onPrev={this.toPrevPage}
            hasNext={this.props.hasNext}
            hasPrev={this.props.hasPrev}
          />
        </div>
      </div>

    );
  }
}

/**
 * Function maps component props to the application state.
 * @param {object} state - application state.
 */
function mapStateToProps(state) {
  return {
    bucketlists: state.bucketlists.items,
    auth: state.auth,
    pages: state.bucketlists.totalpages,
    current: state.bucketlists.currentpage,
    next: state.bucketlists.nextpage,
    prev: state.bucketlists.prevpage,
    hasNext: state.bucketlists.has_next,
    hasPrev: state.bucketlists.has_prev
   };
}
// Connect component to redux store.
export default connect(
  mapStateToProps,
  { getBucketlists, deleteBucketlist, clearMessages, searchBucketlists }
)(BucketLists);
