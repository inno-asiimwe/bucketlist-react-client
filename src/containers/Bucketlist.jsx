import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import _ from 'lodash';
import { getBucketlist, deleteBucketlistItem } from '../actions/action_bucketlist';

export class ShowBucketlist extends Component {
  constructor(props) {
    super(props);
    this.onDelete = this.onDelete.bind(this);
  }
  componentDidMount() {
    const { id } = this.props.match.params;
    this.props.getBucketlist(id);
  }
  onDelete(bucketlistId, itemId) {
    this.props.deleteBucketlistItem(
      bucketlistId,
      itemId,
      () => this.props.getBucketlist(bucketlistId)
    );
  }
  renderItems(id) {
    return _.map(this.props.bucketlist.items, item => (
      <tr>
        <td>
          {item.name}
        </td>
        <td>
          <Link className="btn btn-primary" to={`/bucketlists/${id}/${item.id}/edit`}>
                Edit
          </Link>
        </td>
        <td>
          <button className="btn btn-danger" onClick={() => this.onDelete(id, item.id)}>
                Delete
          </button>
        </td>
      </tr>
    ));
  }
  render() {
    const { bucketlist } = this.props;

    if (!this.props.auth.Authenticated) {
      return <Redirect to="/login" />;
    }

    if (!bucketlist) {
      return <div>Loading ...</div>;
    }
    const { name } = bucketlist;
    const { description } = bucketlist;
    const { id } = bucketlist;
    return (
      <div className="container">
        <div className="jumbotron">
          <div>
            <div className="float-left">
              <Link className="btn btn-primary" to="/" >Back to Bucketlists</Link>
            </div>
            <div className="float-right text-right">
              <Link className="btn btn-primary" to={`/bucketlists/${id}/new`} >Add Bucketlist item</Link>
            </div>
          </div>
          <div>
            <h3>{name} </h3>
            <p>description: {description}</p>
            <h4>{`items in ${name}`}</h4>
            <table className="table">
              <tbody>
                {this.renderItems(id)}
              </tbody>
            </table>
          </div>
        </div>
      </div>
    );
  }
}
function mapStateToProps(state, ownProps) {
  return { bucketlist: state.bucketlists[ownProps.match.params.id], auth: state.auth };
}
export default connect(mapStateToProps, { getBucketlist, deleteBucketlistItem })(ShowBucketlist);
