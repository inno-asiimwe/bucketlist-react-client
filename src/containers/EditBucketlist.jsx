import React, { Component } from 'react';
import { connect } from 'react-redux';

import { getBucketlist, editBucketlist } from '../actions/action_bucketlist';
import UpdateBucketlistForm from '../components/UpdateBucketList';


class EditBucketlist extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }
  componentDidMount() {
    const { id } = this.props.match.params;
    this.props.getBucketlist(id);
  }
  onSubmit(values) {
    const { id } = this.props.match.params;
    this.props.editBucketlist(id, values, () => {
      this.props.history.push('/');
    });
  }
  render() {
    const { bucketlist } = this.props;
    if (!bucketlist) {
      return <div> Loading...</div>;
    }
    const { name } = this.props.bucketlist;
    const { description } = this.props.bucketlist;
    return (
      <div>
           Edit bucketlist
           <UpdateBucketlistForm
             initialValues={{ name, description }}
             onSubmit={this.onSubmit}
             entity="Bucketlist"
             returnTo="/"
           />
      </div>
    );
  }
}

function mapStateToProps(state, ownProps) {
  return { bucketlist: state.bucketlists[ownProps.match.params.id] };
}

export default connect(mapStateToProps, { getBucketlist, editBucketlist })(EditBucketlist);
