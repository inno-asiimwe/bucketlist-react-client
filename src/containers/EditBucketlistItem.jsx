import React, { Component } from 'react';
import { connect } from 'react-redux';

import { getBucketlist, editItem } from '../actions/action_bucketlist';
import UpdateBucketlistForm from './UpdateBucketList';


class EditBucketlistItem extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }
  componentDidMount() {
    const { id } = this.props.match.params;
    this.props.getBucketlist(id);
  }
  onSubmit(values) {
    const { id } = this.props.match.params;
    const { itemid } = this.props.match.params;
    this.props.editItem(id, itemid, values, () => {
      this.props.history.push(`/bucketlists/${id}`);
    });
  }
  render() {
    const { item } = this.props;
    if (!item) {
      return <div> Loading...</div>;
    }
    const { name } = this.props.item;
    const { description } = this.props.item;
    return (
      <div>
           Edit bucketlist Item
           <UpdateBucketlistForm
             initialValues={{ name, description }}
             onSubmit={this.onSubmit}
             entity="Item"
             returnTo={`/bucketlists/${this.props.match.params.id}`}
           />
      </div>
    );
  }
}

function mapStateToProps(state, ownProps) {
  return { item: state.bucketlists[ownProps.match.params.id].items[ownProps.match.params.itemid] };
}

export default connect(mapStateToProps, { getBucketlist, editItem })(EditBucketlistItem);
