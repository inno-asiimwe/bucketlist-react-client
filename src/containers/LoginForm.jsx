/**
 * Component handles login 
 */
import React, { Component } from 'react';
import _ from 'lodash';
import { Field, reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import { Link, Redirect } from 'react-router-dom';
import { ToastContainer, toast } from 'react-toastify';
import { loginUser, clearMessages } from '../actions';

const FIELDS = ['username', 'password'];

/**
 * Function renders the fields in a form
 * @param {object} field - field for the form
 */
function renderField(field) {
  const { meta: { touched, error } } = field;
  const className = `form-group row ${touched && error ? 'has-danger' : ''}`;
  return (
    <div className={className}>
      <label>{field.label}</label>
      <input
        className="form-control"
        type={field.type}
        {...field.input}
      />
      <div className="form-control-feedback">
        {touched ? error : ''}
      </div>
    </div>
  );
}
export class LoginUser extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }

  /**
   * Function executed when form is submitted
   * @param {object} values - username and password
   */
  onSubmit(values) {
    // Dispatch action to login user with submitted credentials
    this.props.loginUser(values);
  }

  /**
   * Function displays an error toast
   * @param {string} - message to toast 
   */
  notify_error = (msg) => {
    toast.error(msg);
  }

  /**
   * Function displays a success toast
   * @param {string} - message to toast
   */
  notify_success = (msg) => {
    toast.success(msg)
  }

  render() {
    // Toast a success message on succesful login
    if (this.props.auth.success && this.props.auth.Authenticated) {
      this.notify_success(this.props.auth.success_msg)
      this.props.clearMessages()
    }
    // Redirect to home page if user is authenticated
    if (this.props.auth.Authenticated ) {
      return <Redirect to="/" />;
    } 
    // Toasr an error message incase of a failure
    else if (this.props.auth.error) {
      this.notify_error(this.props.auth.error_msg)
      this.props.clearMessages();
    }
    const { handleSubmit } = this.props;
    return (
      <div className="container">
        <ToastContainer />
        <h3> Login </h3>
        <br />
        <form onSubmit={handleSubmit(this.onSubmit)}>

          <Field
            label="Username"
            name="username"
            typoe="input"
            component={renderField}
          />
          <Field
            label="Password"
            name="password"
            type="password"
            component={renderField}
          />
          <button type="submit" className="btn btn-primary">Submit</button>
          <Link to="/" className="btn btn-danger">Cancel</Link>
        </form>
      </div>
    );
  }
}

function validate(values) {
  const errors = {};
  // validating the inputs from 'values'
  _.forEach(FIELDS, (field) => {
    if (!values[field]) {
      errors[field] = `Enter a ${field}`;
    }
  });
  return errors;
}
/**
 * 
 * @param {object} state - application state
 */
function mapStateToProps(state) {
  return { auth: state.auth };
}
// connect the redux form to the redux store 
export default reduxForm({
  validate,
  form: 'LoginForm'
})(connect(mapStateToProps, { loginUser, clearMessages })(LoginUser));
