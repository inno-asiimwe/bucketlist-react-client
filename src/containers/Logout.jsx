import React, { Component } from 'react';
import { connect } from 'react-redux';
import { logoutUser } from '../actions/action_auth';


class Logout extends Component {
  componentWillMount() {
    this.props.logoutUser(() => { this.props.history.push('/login'); });
  }
  render() {
    return null;
  }
}

export default connect(null, { logoutUser })(Logout);
