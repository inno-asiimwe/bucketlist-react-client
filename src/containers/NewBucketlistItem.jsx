import React, { Component } from 'react';
import _ from 'lodash';
import { Field, reduxForm } from 'redux-form';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { addBucketlistItem } from '../actions';

const FIELDS = ['name', 'description'];
function renderField(field) {
  const { meta: { touched, error } } = field;
  const className = `form-group row ${touched && error ? 'has-danger' : ''}`;
  return (
    <div className={className}>
      <label>{field.label}</label>
      <input
        className="form-control"
        type={field.type}
        {...field.input}
      />
      <div className="form-control-feedback">
        {touched ? error : ''}
      </div>
    </div>
  );
}
class NewItem extends Component {
  constructor(props) {
    super(props);
    this.onSubmit = this.onSubmit.bind(this);
  }
  onSubmit(values) {
    const { id } = this.props.match.params;
    this.props.addBucketlistItem(id, values, () => {
      this.props.history.push(`/bucketlists/${id}`);
    });
  }
  render() {
    const { handleSubmit } = this.props;
    const { id } = this.props.match.params;
    return (
      <div className="container">
        <h3> New Bucketlist Item</h3>
        <br />
        <form onSubmit={handleSubmit(this.onSubmit)}>

          <Field
            label="Item Name"
            name="name"
            type="input"
            component={renderField}
          />
          <Field
            label="Item Description"
            name="description"
            type="text"
            component={renderField}
          />
          <button type="submit" className="btn btn-success">Create</button>
          <Link to={`/bucketlists/${id}`} className="btn btn-danger">Cancel</Link>
        </form>
      </div>
    );
  }
}

function validate(values) {
  const errors = {};
  // validating the inputs from 'values'
  _.forEach(FIELDS, (field) => {
    if (!values[field]) {
      errors[field] = `Enter a ${field}`;
    }
  });
  return errors;
}

export default reduxForm({
  validate,
  form: 'NewItemForm'
})(connect(null, { addBucketlistItem })(NewItem));
