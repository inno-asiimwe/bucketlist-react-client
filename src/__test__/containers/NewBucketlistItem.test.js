import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import { NewItem } from '../../containers/NewBucketlistItem';

describe('NewItem', () => {
  it('Matches its snapshot', () => {
    const output = shallow(<NewItem
      match={{ params: { id: 1 } }}
      handleSubmit={() => {}}
      auth={{ Authenticated: true }}
    />);
    expect(shallowToJson(output)).toMatchSnapshot();
  });
});
