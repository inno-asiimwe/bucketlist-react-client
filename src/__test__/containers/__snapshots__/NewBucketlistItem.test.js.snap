// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewItem Matches its snapshot 1`] = `
<div>
  Loading...
</div>
`;
